src/devices/GoPiGo3/Movements/Vehicle.cs

```diff
@@ -0,0 +1,270 @@
+
+        /// <summary>
+        /// Is the vehicle has inverted direction, then true
+        /// </summary>
+        public bool DirectionOpposite
+        {
+            get { return _directionOpposite; }
+
+            set
+            {
+                _directionOpposite = value;
+                _correctedDir = (_directionOpposite) ? -1 : 1;
+            }
+        }
+
+        private void RunMotorSyncTime(MotorPort[] ports, int[] speeds, int timeout)
+        {
+            if ((ports == null) || (speeds == null))
+                return;
+            if (ports.Length != speeds.Length)
+                return;
+            //create a timer for the needed time to run
+            if (_timer == null)
+                _timer = new Timer(RunUntil, null, TimeSpan.FromMilliseconds(timeout), Timeout.InfiniteTimeSpan);
+            else
+                _timer.Change(TimeSpan.FromMilliseconds(timeout), Timeout.InfiniteTimeSpan);
+
+            //initialize the speed and enable motors
+            for (int i = 0; i < ports.Length; i++)
+            {
+                StartMotor(ports[i], speeds[i]);
+            }
+
+            bool nonstop = true;
+            while (nonstop)
+            {
+                bool status = false;
+                for (int i = 0; i < ports.Length; i++)
+                {
+                    status |= IsRunning(ports[i]);
+                }
+                nonstop = status;
+            }
+        }
+
+        private void RunUntil(object state)
+        {
+            StopMotor(PortLeft);
+            StopMotor(PortRight);
+            if (_timer != null)
+            {
+                _timer.Dispose();
+                _timer = null;
+            }
+        }
+
+        private void StopMotor(MotorPort port)
+        {
+            _goPiGo.SetMotorPower(port, 0);
+        }
+
+        private void StartMotor(MotorPort port, int speed)
+        {
+            speed = Math.Clamp(speed, -255, 255);
+            _goPiGo.SetMotorPower(port, speed);
+        }
+
+        private void RunMotorSyncDegrees(MotorPort[] ports, int[] speeds, int[] degrees)
+        {
+            if ((ports == null) || (speeds == null) || degrees == null)
+                return;
+            if ((ports.Length != speeds.Length) && (degrees.Length != speeds.Length))
+                return;
+            //make sure we have only positive degrees
+            for (int i = 0; i < degrees.Length; i++)
+            {
+                if (degrees[i] < 0)
+                    degrees[i] = -degrees[i];
+            }
+
+            //initialize the speed and enable motors
+            int[] initval = new int[ports.Length];
+            for (int i = 0; i < ports.Length; i++)
+            {
+                try
+                {
+                    initval[i] = _goPiGo.GetMotorEncoder(ports[i]);
+                    StartMotor(ports[i], speeds[i]);
+                }
+                catch (IOException)
```

Comment:
I don't understand this pattern. Why are we throwing an exception in `GetMotorEncoder` if we don't want callers to do something when an invalid port is passed in?